src/NLog/GlobalDiagnosticsContext.cs

```diff
@@ -152,10 +142,40 @@ public static void Remove(string item)
         /// </summary>
         public static void Clear()
         {
-            lock (dict)
+            lock (_dict)
+            {
+                GetWritableDict(_dictReadOnly != null && _dict.Count > 0, false).Clear();
+            }
+        }
+
+        private static Dictionary<string, object> GetReadOnlyDict()
+        {
+            var readOnly = _dictReadOnly;
+            if (readOnly == null)
+            {
+                lock (_dict)
+                {
+                    readOnly = _dictReadOnly = _dict;
+                }
+            }
+            return readOnly;
+        }
+
+        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clone = true)
+        {
+            if (mustCreateNew)
             {
-                dict.Clear();
+                var newDict = new Dictionary<string, object>(clone ? _dict.Count + 1 : 0);
```

Comment:
Is it correct that clear has the init capacity of 0, and the field has the default capacity (16?). It that on purpose?

Reply:
Don't understand your question. But Clear (clone=false) calls with capacity = 0 that means do not allocate internal arrays until insert happens.

Reply:
indeed, but the init doesn't set it to `0`

`private static Dictionary<string, object> _dict = new Dictionary<string, object>();`

Reply:
Still don't understand your queustion. The default constructor for Dictionary uses capacity = 0. See also https://referencesource.microsoft.com/#mscorlib/system/collections/generic/dictionary.cs,85